python3Packages.torch: Suppress spammy warnings

Madouura opened the pull request:
## Description of changes
Fixes #266895
~~Before it is asked, we cannot just simply define the ``HIP_CXX_FLAGS`` variable in ``preConfigure``.~~

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Artturin commented on pkgs/development/python-modules/torch/default.nix:
```diff
@@ -150,6 +150,8 @@ in buildPythonPackage rec {
     # https://github.com/pytorch/pytorch/pull/108847
     ./pytorch-pr-108847.patch
+  ] ++ lib.optionals rocmSupport [
+    ./0000-rocm-suppress-spammy-warnings.patch
```
The build system can be bypassed with

`NIX_CFLAGS_COMPILE`

ex: https://github.com/NixOS/nixpkgs/blob/bab997db72a94fd58c6d3864f71ad8a81544cc6a/pkgs/development/libraries/libdynd/default.nix#L18

There are a lot of warnings even without `rocmSupport` so it would be good to reduce those too.

Madouura replied:
Thanks for pointing that out!

Madouura committed:
python3Packages.torch: Suppress spammy warnings

Suppresses the most spammy warnings.
This is mainly to fix https://github.com/NixOS/nixpkgs/issues/266895.

Madouura commented:
Suppressed the worst of the warnings for regular and ROCm variants.
Not doing CUDA because it's not cached anyway.
``--amdgpu-target`` warning spam will be fixed by #259068.

tpwrules commented:
Would be a good idea to add a comment explaining why the flags were added.

wegank commented:
Could you please rebase this on staging-next? The flags there are different for Clang support.

Madouura commented:
> Would be a good idea to add a comment explaining why the flags were added.

The comment is in the commit message. @mweinelt keyed me in on why this is better than adding it to the derivation.

> Could you please rebase this on staging-next? The flags there are different for Clang support.

I'll try building it on staging-next locally, but if it works the same I don't think there's a reason to rebase here on the PR?

Artturin commented on pkgs/development/python-modules/torch/default.nix:
```diff
@@ -290,6 +290,22 @@ in buildPythonPackage rec {
+    "-Wno-unused-command-line-argument"
+    "-Wno-maybe-uninitialized"
+    "-Wno-uninitialized"
+    "-Wno-array-bounds"
```
```suggestion
    "-Wno-stringop-overflow"
```

Artturin replied:
Is there a `-Wall` somewhere in the code?

Artturin replied:
This in postPatch
`find . -type f -iname "*.cmake" -o -iname "CMakeLists.txt" -o -iname "Makefile" -exec sed -i '/-Wall/d' {} \;`

will get rid of the `Wall`'s but many of these projects enable individual `-W` too so it's not very useful

Madouura replied:
I only got rid of the absolute worst spammy warnings, I see no reason to get rid of them all.

Madouura replied:
Also I think getting rid of all the warning options could be prone to breakage, so best avoided.

wegank commented:
Has anyone tested this on darwin before merging? It seems that some flags are not recognized by Clang in #267209.
